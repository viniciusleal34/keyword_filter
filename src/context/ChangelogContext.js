/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-no-constructed-context-values */

import React, { createContext, useState } from 'react';
import { searchLogAsync } from '../services/api-service';
import TypeComponent from '../utils/getLabelTypeButton';

const ChangelogContext = createContext();

const ChangelogStore = ({ children }) => {
    const [logData, setLogData] = useState([]);
    const [informationPage, setInformationPage] = useState({});
    const [loadingItems, setLoadingItems] = useState(true);
    const [search, setSearch] = useState('');

    const searchLogs = async (page) => {
        const resp = await searchLogAsync(search, page);
        // delete resp?.logs;
        // setInformationPage(resp);
        return resp;
    };

    const constructObjectsData = async (response, isSearch = false) => {
        setLoadingItems(true);
        const logObjects = response?.logs?.map((logObject) => ({
            user: logObject.user,
            keyword: logObject.keyword,
            editDate: logObject.editDate,
            action: logObject.action,
            componentType: TypeComponent[logObject.componentType],
            actualValue: JSON.parse(logObject.actualValue),
            oldValue: JSON.parse(logObject.oldValue)
        }));
        if (isSearch) {
            setLogData(logObjects);
        } else {
            setLogData(() => [...logData, ...logObjects]);
        }
        delete response?.logs;
        setInformationPage(response);
        setTimeout(() => {
            setLoadingItems(false);
        }, 300);
    };

    return (
        <ChangelogContext.Provider
            value={{
                logData,
                setLogData,
                searchLogs,
                informationPage,
                setInformationPage,
                loadingItems,
                constructObjectsData,
                setLoadingItems,
                setSearch,
                search
            }}
        >
            {children}
        </ChangelogContext.Provider>
    );
};

export { ChangelogContext, ChangelogStore };
