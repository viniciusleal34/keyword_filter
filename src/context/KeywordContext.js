/* eslint-disable no-unused-vars */
/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-no-constructed-context-values */
import React, { createContext, useEffect, useState } from 'react';
import ProjectString from '../constants/project-constants';
import {
    deleteKeywordJsonAsync,
    deleteKeywordLinkAsync,
    deleteKeywordRedirectAsync,
    deleteKeywordTextAsync,
    getKeywordOneAsync,
    getListKeywordAsync,
    postKeywordAsync,
    postKeywordJsonAsync,
    postKeywordLinkAsync,
    postKeywordRedirectAsync,
    postKeywordTextAsync,
    putHeaderKeywordAsync,
    putKeywordAsync,
    putKeywordJsonAsync,
    putKeywordLinkAsync,
    putKeywordRedirectAsync,
    putKeywordTextAsync,
    searchKeyWordAsync
} from '../services/api-service';
import { showToast } from '../services/common-service';

export const KeywordContext = createContext();

export const KeywordStorage = ({ children }) => {
    const [buttonArray, setButtonArray] = useState([]);
    const [keywordList, setKeywordList] = useState([]);
    const [loadingItems, setLoadingItems] = useState(true);
    const [informationPage, setInformationPage] = useState({});
    const [search, setSearch] = useState('');
    const [initialDataKeyword, setInitialDataKeyword] = useState({
        keyword: '',
        title: '',
        menuMessage: '',
        id: ''
    });
    const [toast, setToast] = useState(false);
    const [openToast, setOpenToast] = useState(false);
    const [isEdit, setIsEdit] = useState(true);

    useEffect(() => {
        setTimeout(() => {
            setOpenToast(false);
            setToast(false);
        }, 2000);
    }, [openToast]);

    const constructKeywordData = async (response, isSearch = false) => {
        setLoadingItems(true);
        const { keywordObjects } = response;
        if (isSearch) {
            setKeywordList(() => [...keywordObjects]);
        } else {
            setKeywordList(() => [...keywordList, ...keywordObjects]);
        }
        delete response?.keywordObjects;
        setInformationPage(response);

        setTimeout(() => {
            setLoadingItems(false);
        }, 300);
    };

    const keywordCreate = async (values, setIsButtons, resetForm, setOpen) => {
        const textButtons = [];
        const linkButtons = [];
        const redirectButtons = [];
        const jsonButtons = [];
        buttonArray?.forEach(async (element) => {
            if (element?.type === ProjectString.TYPE_TEXT) {
                textButtons.push(element);
            } else if (element?.type === ProjectString.TYPE_LINK) {
                linkButtons.push(element);
            } else if (element?.type === ProjectString.TYPE_REDIRECT) {
                redirectButtons.push(element);
            } else if (element?.type === ProjectString.TYPE_JSON) {
                jsonButtons.push(element);
            }
        });
        const request = {
            ...values,
            isTextHeader: true,
            textButtons,
            linkButtons,
            jsonButtons,
            redirectButtons
        };
        const response = await postKeywordAsync(request);
        console.log(response);

        if (response === false) {
            return false;
        }
        const resp = await getListKeywordAsync();
        const { keywordObjects } = resp;
        setKeywordList(keywordObjects);
        setOpen(false);
        setIsButtons(false);
        resetForm({});
        setButtonArray([]);
        await showToast({
            type: 'success',
            message: 'Adicionado com sucesso !'
        });
        return response;
    };

    const keywordUpdate = async (newValue, setOpen) => {
        let response = false;
        if (initialDataKeyword?.keyword !== newValue?.keyword) {
            response = await putKeywordAsync(newValue);
        }
        newValue.id = initialDataKeyword?.headerId;
        response = await putHeaderKeywordAsync({
            ...newValue,
            isTextHeader: true,
            ImageUri: null
        });

        if (response !== false) {
            const resp = await getListKeywordAsync();
            const { keywordObjects } = resp;
            setKeywordList(keywordObjects);
            setOpen(false);
            await showToast({
                type: 'success',
                message: 'Alterado com sucesso!'
            });
        }
        return response;
    };

    const buttonKeywordCreate = async (values, resetForm) => {
        values.KeywordMenuId = initialDataKeyword?.id;
        let response = null;
        if (values?.type === ProjectString.TYPE_TEXT) {
            response = await postKeywordTextAsync(values);
        }
        if (values?.type === ProjectString.TYPE_LINK) {
            response = await postKeywordLinkAsync(values);
        }
        if (values?.type === ProjectString.TYPE_REDIRECT) {
            response = await postKeywordRedirectAsync(values);
        }
        if (values?.type === ProjectString.TYPE_JSON) {
            response = await postKeywordJsonAsync(values);
        }

        if (response !== false) {
            const resp = await getListKeywordAsync();
            const { keywordObjects } = resp;
            setKeywordList(keywordObjects);
            restoreData(response?.id);
            resetForm({});
            await showToast({
                type: 'success',
                message: 'Adicionado com sucesso!'
            });
        }

        return response;
    };

    const buttonKeywordUpdate = async (values, setEditButton) => {
        let response = null;
        if (values?.type === ProjectString.TYPE_TEXT) {
            response = await putKeywordTextAsync(values);
        }
        if (values?.type === ProjectString.TYPE_LINK) {
            response = await putKeywordLinkAsync(values);
        }
        if (values?.type === ProjectString.TYPE_REDIRECT) {
            response = await putKeywordRedirectAsync(values);
        }
        if (values?.type === ProjectString.TYPE_JSON) {
            response = await putKeywordJsonAsync(values);
        }
        if (response === false) {
            return false;
        }
        const resp = await getListKeywordAsync();
        const { keywordObjects } = resp;
        setKeywordList(keywordObjects);
        restoreData(response?.id);
        setEditButton(false);
        await showToast({
            type: 'success',
            message: 'Atualizado com sucesso!'
        });
        return response;
    };

    const buttonKeywordDelete = async (values, newArray) => {
        let response = null;
        if (values?.type === ProjectString.TYPE_TEXT) {
            response = await deleteKeywordTextAsync(values?.id);
        }
        if (values?.type === ProjectString.TYPE_LINK) {
            response = await deleteKeywordLinkAsync(values?.id);
        }
        if (values?.type === ProjectString.TYPE_REDIRECT) {
            response = await deleteKeywordRedirectAsync(values?.id);
        }
        if (values?.type === ProjectString.TYPE_JSON) {
            response = await deleteKeywordJsonAsync(values?.id);
        }
        if (response !== false) {
            const resp = await getListKeywordAsync();
            const { keywordObjects } = resp;
            setKeywordList(keywordObjects);
            setButtonArray(() => [...newArray]);
            await showToast({
                type: 'success',
                message: 'Deletado com sucesso!'
            });
        }

        return false;
    };

    const previewComponent = async (values) => {
        const request = {
            ...values
        };
        return request;
    };

    const orderButtonArray = (response) => {
        const buttons = [];
        response?.textButtons?.forEach((button) => {
            button.type = ProjectString.TYPE_TEXT;
            buttons.push(button);
        });
        response?.linkButtons?.forEach((button) => {
            button.type = ProjectString.TYPE_LINK;
            buttons.push(button);
        });
        response?.redirectButtons?.forEach((button) => {
            button.type = ProjectString.TYPE_REDIRECT;
            buttons.push(button);
        });
        response?.jsonButtons?.forEach((button) => {
            button.type = ProjectString.TYPE_JSON;
            button.value = JSON.stringify(button.value);
            buttons.push(button);
        });
        const buttonOrder = buttons.sort((a, b) => {
            if (a.position < b.position) {
                return -1;
            }
            return 0;
        });
        setButtonArray(() => [...buttonOrder]);
    };

    const restoreData = async (id) => {
        const response = await getKeywordOneAsync(id);
        if (!response?.errorMessage) {
            setInitialDataKeyword({
                id: response?.id,
                keyword: response?.keyword,
                title: response?.header?.title,
                menuMessage: response?.header?.menuMessage,
                headerId: response?.header?.id
            });
            orderButtonArray(response);
            setIsEdit(true);
            return true;
        }
        setIsEdit(false);
        return response;
    };

    const searchKeyword = async (keyword, page = 1) => {
        if (keyword === '') {
            const resp = await getListKeywordAsync();
            return resp;
        }
        const response = await searchKeyWordAsync(keyword, page);
        return response;
    };

    return (
        <KeywordContext.Provider
            value={{
                setToast,
                toast,
                openToast,
                setOpenToast,
                keywordCreate,
                buttonArray,
                setButtonArray,
                previewComponent,
                initialDataKeyword,
                setInitialDataKeyword,
                restoreData,
                setIsEdit,
                isEdit,
                buttonKeywordCreate,
                buttonKeywordUpdate,
                buttonKeywordDelete,
                keywordUpdate,
                setKeywordList,
                keywordList,
                searchKeyword,
                setSearch,
                search,
                informationPage,
                setInformationPage,
                setLoadingItems,
                loadingItems,
                constructKeywordData
            }}
        >
            {children}
        </KeywordContext.Provider>
    );
};
