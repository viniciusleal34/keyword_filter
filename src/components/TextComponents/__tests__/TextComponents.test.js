/* eslint-disable react/jsx-no-constructed-context-values */
/* eslint-disable no-unused-vars */

import React from 'react';
import { render, screen } from '@testing-library/react';
import { KeywordContext } from '../../../context/KeywordContext';

import { TextComponents } from '../index';

const buttonArray = [
    {
        id: '0e08041f-1cde-41ac-8b31-284ccdde7cea',
        title: 'teste1',
        value: 'teste1',
        type: 'text'
    },
    {
        id: 'ff14bc59-12aa-45d0-90f7-0f9b2f5a917b',
        title: 'teste2',
        value: 'teste2',
        type: 'text'
    },
    {
        id: '71a433bb-d153-4ed1-865e-fe472eb8b935',
        title: 'teste3',
        uri: 'https://portal.blip.ai/application/detail/plugin3/plugin/keyword-filter',
        type: 'link'
    },
    {
        id: 'fd1317ba-91d9-4103-b587-0dc40de691c8',
        title: 'das',
        destinationService: 'teste4',
        destinationState: 'teste6',
        type: 'redirect'
    }
];
const buttonObject = {
    headerId: '3a1b0ace-1512-4832-b3a2-fda427e950fa',
    id: '478ea40d-be91-4968-a57b-41400589b681',
    keyword: 'keyword',
    menuMessage: 'lorem Lorem lorem',
    title: 'Palavra'
};

describe('Testing with Component is visible', () => {
    const handleClick = jest.fn();
    let rtlContainer;

    beforeEach(() => {
        const { container } = render(
            <KeywordContext.Provider value={{ buttonArray }}>
                <TextComponents
                    buttonObject={buttonObject}
                    open
                    setOpen={handleClick}
                />
            </KeywordContext.Provider>
        );
        rtlContainer = container;
    });

    it('Should modal is open', () => {
        expect(screen.getByText('teste1')).toBeTruthy();
        expect(screen.getByText('teste2')).toBeTruthy();
        expect(screen.getByText('Palavra')).toBeTruthy();
        expect(screen.getByText('lorem Lorem lorem')).toBeTruthy();
    });
});

describe('Testing with Component is not visible', () => {
    const handleClick = jest.fn();
    let rtlContainer;

    beforeEach(() => {
        const { container } = render(
            <KeywordContext.Provider value={{ buttonArray }}>
                <TextComponents
                    buttonObject={buttonObject}
                    open={false}
                    setOpen={handleClick}
                />
            </KeywordContext.Provider>
        );
        rtlContainer = container;
    });

    it('should snapshot',() =>{
        expect(rtlContainer.firstChild).toMatchSnapshot();
    });

    it('Should modal is close', () => {
        expect(screen.queryByText('lorem Lorem lorem')).not.toBeTruthy();
    });
    
});
