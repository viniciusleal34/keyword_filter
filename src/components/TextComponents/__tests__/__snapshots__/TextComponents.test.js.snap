// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing with Component is not visible should snapshot 1`] = `null`;
